KEP 3589: Uniformly filter manageJobsWithoutQueueNames by namespace (#3595)
kubernetes-sigs/kueue: keps/3589-manage-jobs-selectively/README.md

The irregularity between `batchv1/Job` and `Deployment` is likely to be especially confusing
to users and cluster admins who would view both of these types as "built-in" to Kubernetes and
would expect them to be configured in the same way.

### Goals

Provide a simple and uniform API to enable cluster admins to configure Kueue
such that quotas will be enforced for specific namespaces.

### Non-Goals

Consider any scope for filtering `manageJobsWithoutQueueName` that is not based on namespaces.

## Proposal

We add a `manageAllJobsNamespaceSelector` of type `*metav1.LabelSelector`
to the top level of the Kueue `Configuration` struct (same level as `manageJobsWithoutQueueName`).
(The name of `manageJobsWithoutQueueNameNamespaceSelector` reads poorly, so I propose something else)

We use this new configuration for all integrations in the same way that the current
`namespaceSelector` is used by the Pod integration.

We migrate away from using `podOptions.namespaceSelector` over the course of several releases.
Ideally we remove it from the `Configuration` struct as part of going to the `v1` API level.

### User Stories (Optional)

#### Story 1

> Cluster admins will be able to use `manageAllJobsNamespaceSelector` to easily configure Kueue

Reviewer: Reading this, I don't quite understand the interaction of ManageJobsWithoutQueueName and ManageAllJobsNamespaceSelector. Could you give an example of how you would configure this and how it would work?

I'm just not entirely sure what someone would put for a label selector in this case.

Author: I expanded the proposal and user story to address this feedback.

	ManageAllJobsNamespaceSelector *metav1.LabelSelector `json:"manageAllJobsNamespaceSelector,omitempty"`

  ...
}
```

### Implementation

The configuration of all Job webhooks are extended to propagate the label selector and it is
used to filter manageJobsWithoutQueueNames as is currently done in the Pod integration.

Configuration parsing during startup is extended to give a deprecation notice if `podOptions.namespaceSelector` is set.
Setting both `podOptions.namespaceSelector` and `manageAllJobsNamespaceSelector` will be flagged as a configuration error.

### Test Plan

[ X ] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

#### Unit Tests

Unit tests for configuration parsing will be added.

All unit tests for the current Pod integration namespaceSelector will be extended across all integrations.

#### Integration tests

All integration tests for the current Pod integration namespaceSelector will be extended across all integrations.

### Graduation Criteria

> The feature is implemented as described for all Job integrations.

Reviewer: I don't think this really answers the question for graduation critiera.

This would usually be what would be required to "graduate" this feature to GA.

I'm not sure if this needs a beta stage..

Author: I dropped this section.

  - [Validating Admission Policies](#validating-admission-policies)
<!-- /toc -->

## Summary

This KEP adds an optional field to the Kueue `Configuration` struct that enables
the effect of `manageJobsWithoutQueueName` to be controlled on a per-namespace
basis for all Kueue integrations.

-->

## Motivation

When Kueue is deployed on multi-tenant clusters, it is essential that cluster
admins be able to configure Kueue such that users cannot bypass its quota system.
In particular, there must be a robust and recommended mechanism for configuring Kueue
such that all Pod-creating Kinds in user namespaces are subject to management by Kueue.

Currently, the primary mechanism for doing this is setting `manageJobsWithoutQueueName` to
true.  When `manageJobsWithoutQueueName` is true, then Kueue will manage all instances of
Kinds with Kueue integrations, even if they do not have a `kueue.x-k8s.io/queue-name` label.
As a practical matter, Pods running in system namespaces must be exempted from being
managed by Kueue even if `manageJobsWithoutQueueName` is true.  Therefore Kueue's Pod
integration limits the scope of `manageJobsWithoutQueueName` by restricting its effect to

> Pods that are in namespaces that match against the `integrations.podOptions.namespaceSelector`.

Reviewer 2: I think that we can just mark podNamespaceSelector as deprecated.

> Pods that are in namespaces that match against the `integrations.podOptions.namespaceSelector`.

Reviewer 3: If we move forward with this KEP we can. Otherwise we still need it to say that pod integration, deployement, and statefulSet is constrained.

The recently added `Deployment` and `StatefulSet` integrations are built on top of the
`Pod` integration and therefore the effect of `manageJobsWithoutQueueName` for these
two integrations is also filtered by `integrations.podOptions.namespaceSelector`.  A similar
rationale holds for `Deployment` and less strongly for `StatefulSet` that some mechanism
to limit `manageJobsWithoutQueueName` is required otherwise it would be impossible in
practice to ever set `manageJobsWithoutQueueName` to true.

No other integrations beyond these three consider the `namespaceSelector` when deciding whether or not to
suspend jobs that do not have a `queue-name` label.  This results in an irregular API and makes it
harder to configure Kueue such that quotas will be enforced (cannot be bypassed by users).
The irregularity between `batchv1/Job` and `Deployment` is likely to be especially confusing
to users and cluster admins who would view both of these types as "built-in" to Kubernetes and
would expect them to be configured in the same way.

### Goals

Provide a simple and uniform API to enable cluster admins to configure Kueue
such that quotas will be enforced for specific namespaces.

### Non-Goals

> Consider any scope for filtering `manageJobsWithoutQueueName` that is not based on namespaces.

Reviewer 2: This is just comment.

My primary concern for this KEP was this one.
I would not like to accept this feature in any case since this tries to re-implement the Gatekeeper, VAP, and MAP.

Reviewer 3: Yes, I fully agree we don't want this mechanism to be more complex, and +1 to reject features not based on namespaces, but for namespaces we already have the complication in `podOptions.namespaceSelector`, and making it uniformly supported seems like good idea when we have Deployments and StatefulSets supported.

I agree with the direction to use VAP and MAP, but this will probably take many releases as MAP is still alpha in 1.32, and we need to propose details for VAP and MAP generation to avoid issues with hierarchical jobs (as described in the KEP).

Reviewer 2: Thanks for expanding that. If anyone wants to know VAP and MAP introduction details, please refer https://github.com/kubernetes-sigs/kueue/issues/3636 and https://github.com/kubernetes-sigs/kueue/issues/3637.

## Proposal

We add a `managedJobsNamespaceSelector` of type `*metav1.LabelSelector`
to the top level of the Kueue `Configuration` struct (same level as `manageJobsWithoutQueueName`).

We use this new configuration for all integrations in the same way that the current
`namespaceSelector` is used by the Pod integration.  Specifically,
1. If `manageJobsWithoutQueueName` is false, `managedJobsNamespaceSelector` has no effect: Kueue will manage
exactly those instances of supported Kinds that have a `queue-name` label.
2. If `manageJobsWithoutQueueName` is true, then Kueue will (a) manage all instances of supported Kinds
that have a `queue-name` label and (b) will manage all instances of supported Kinds that do not
have a `queue-name` label if they are in namespaces that match `managedJobsNamespaceSelector`.

We migrate away from using `podOptions.namespaceSelector` over the course of several releases.

> Ideally we remove it from the `Configuration` struct as part of going to the `v1` API level.

Reviewer 3: Sgtm, I think we can say that in the transition period the actual namespace selector used for Pod integration (and thus StatefulSet and Deployments) will be the intersection of both `managedJobsNamespaceSelector` and `podOptions.namespaceSelector`.

### User Stories (Optional)

#### Story 1

Cluster admins will be able to use the combination of `managedJobsNamespaceSelector` and
`manageJobsWithoutQueueName` to configure Kueue to enable quota enforcement on a per namespace basis.

> One sapproach would be to exempt selected namespaces from management:

Reviewer 2: ```suggestion
One approach would be to exempt selected namespaces from management:
```

```yaml
managedJobsNamespaceSelector:
  matchExpressions:
  - key: kubernetes.io/metadata.name
    operator: NotIn
    values: [ kube-system, kueue-system ]
```
Another approach would be for the cluster admin to label namespaces that are subject to management:
```yaml
managedJobsNamespaceSelector:
  matchLabels:
    kueue.x-k8s.io/managed-namespace: true
```

## Design Details

### API

> The `Configuration` struct is extended to add ManageAllJobsNamespaceSelector

Reviewer 2: ```suggestion
The `Configuration` struct is extended to add ManagedJobsNamespaceSelector
```

```go
type Configuration struct {

  ...

> 	// ManageJobsWithoutQueueName controls whether or not Kueue reconciles

Reviewer 2: This is just a comment and not an objection to this feature.

I was wondering if we could just deprecate and remove the `manageJobsWithoutQueueName` and introduce more informable validations so that batch users can easily find the quota violation.

I think that the current manageJobsWithoutQueuenName does not have a good user experience since there is no way to find the reason why jobs are not runnable when they accidentally submit Jobs without queue names.

Because the Kueue automatically suspends the Jobs.

However, based on some consideration, this `manageJobsWithoutQueueName` has already been released. So, even if we do not implement this, we must follow the appropriate deprecation steps.

Additionally, if the cluster administrator provides the automatic queue-name insertion mechanism, this could be helpful for the batch users.
Furthermore, such an informable validation mechanism could be an alternative new feature since validation and mutation could not cover all use cases each other, and I think that both are valuable.

	// jobs that don't set the label kueue.x-k8s.io/queue-name.
	// If set to true, then those jobs will be suspended and never started unless
	// they are assigned a queue and eventually admitted. This also applies to
	// jobs created before starting the kueue controller.
	// Defaults to false; therefore, those jobs are not managed and if they are created
	// unsuspended, they will start immediately.
	ManageJobsWithoutQueueName bool `json:"manageJobsWithoutQueueName"`

>   // ManageAllJobsNamespaceSelector can be used to omit some namespaces from ManageJobsWithoutQueueName

Reviewer 3: ```suggestion
  // ManagedJobsNamespaceSelector can be used to omit some namespaces from ManagedJobsWithoutQueueName
```
nit, to align the naming

All unit tests for the current Pod integration namespaceSelector will be extended across all integrations.

#### Integration tests

All integration tests for the current Pod integration namespaceSelector will be extended across all integrations.

## Implementation History

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

### Do nothing

We could do nothing and simply document that despite its historical placement in
a subfield of the `Configuration` struct, `podOptions.namespaceSelector` also applies
to the `StatefulSet` and `Deployment` integrations.

### Validating Admission Policies

As previously discussed in [issue 2119](https://github.com/kubernetes-sigs/kueue/issues/2119),
we could phase out `manageJobsWithoutQueueNames` and instead recommend the
use of ValidatingAdmissionPolicies to ensure that all Kinds that could be managed by Kueue that
are created in "user" namespaces have a queue name label (and thus will be managed by Kueue).

However, while exploring this idea further a complication arose. The ValidatingAdmissionPolicy
must be able to distinguish between top-level resources without queue names (which should be denied)

> and child resources that are owned by Kueue-managed parents (which must be admitted even if they do not have a queue name).

Reviewer 3: Yes, I can understand why VAP becomes complicated when there are long chains of CRDs. I think this picture may be even more complex than in this example if we have multiple parent CRDs which can own batch/v1 Job.

Also, introducing `ManagedJobsNamespaceSelector` isn't really adding new complexity assuming we deprecate and drop podOptions.namespaceSelector over time (another comment)

Author: I think this is the key point.  For either `manageJobsWithoutQueueName` or the default LocalQueue name being pursued in #3610 to be usable, there _has_ to be a way to control which namespaces they apply to for at least the `Pod`, `StatefulSet` and `Deployment` integrations.   On OpenShift, which tends to use `Job` for system operations the namespace control needs to also cover the `Job` integration as well. 

I can rework this KEP to focus on just the generalization of `integrations.podOptions.namespaceSelector` to `managedJobsNamespaceSelector`.    I think we need it independent of what happens to `manageJobsWithoutQueueName`.

```go
type Configuration struct {

  ...

	// ManageJobsWithoutQueueName controls whether or not Kueue reconciles
	// jobs that don't set the label kueue.x-k8s.io/queue-name.
	// If set to true, then those jobs will be suspended and never started unless
	// they are assigned a queue and eventually admitted. This also applies to
	// jobs created before starting the kueue controller.
	// Defaults to false; therefore, those jobs are not managed and if they are created
	// unsuspended, they will start immediately.
	ManageJobsWithoutQueueName bool `json:"manageJobsWithoutQueueName"`

  // ManageAllJobsNamespaceSelector can be used to omit some namespaces from ManageJobsWithoutQueueName
	ManagedJobsNamespaceSelector *metav1.LabelSelector `json:"managedJobsNamespaceSelector,omitempty"`

  ...
}
```

### Implementation

The configuration of all Job webhooks are extended to propagate the namespace selector and it is
used to filter `manageJobsWithoutQueueName` as is currently done in the Pod integration. Specifically,
jobs without queue names are only managed if both `manageJobsWithoutQueueName` is true and
the jobs namespace matches the `managedJobsNamespaceSelector`.

> Configuration parsing during startup is extended to give a deprecation notice if `podOptions.namespaceSelector` is set.

Reviewer 4: One note, the proposed configuration doesn't have a way of impersonating `podOptions.podSelector`. Would we be okay with deprecating it anyways?

Author: I think `podOptions.podSelector` is orthogonal to `podOptions.namespaceSelector`.  This KEP is only proposing to deprecate `podOptions.namespaceSelector` in favor of the new namespaceSelector that would apply to all integrations.

a subfield of the `Configuration` struct, `podOptions.namespaceSelector` also applies
to the `StatefulSet` and `Deployment` integrations.

### Validating Admission Policies

As previously discussed in [issue 2119](https://github.com/kubernetes-sigs/kueue/issues/2119),
we could phase out `manageJobsWithoutQueueNames` and instead recommend the
use of ValidatingAdmissionPolicies to ensure that all Kinds that could be managed by Kueue that
are created in "user" namespaces have a queue name label (and thus will be managed by Kueue).

However, while exploring this idea further a complication arose. The ValidatingAdmissionPolicy
must be able to distinguish between top-level resources without queue names (which should be denied)
and child resources that are owned by Kueue-managed parents (which must be admitted even if they do not have a queue name).
In other words, the VAP must implement the same functionality as `jobframework.IsOwnerManagedByKueue`
and cross check controller-references against managed kinds.  This approach may be challenging to scale
as the number of Kueue managed kinds increases and ancestor trees become deeper.

> For example in the not so distant future we might see: AppWrapper->PyTorchJob->JobSet->Job->Pod.

Reviewer 4: Wouldn't it be enough to do the same as is done in `jobframework.IsOwnerManagedByKueue` and just check if the parent is manageable by Kueue? So ancestor tree size wouldn't matter.

Author: I will try to phrase this better.  It isn't that the check has to actually crawl up the ancestor tree.  It's that the check has to consider each of the kinds that could be Kueue-managed parents.  So it needs a longer list of kinds and it has to somehow know that a particular Kind's Kueue integration is enabled.   For kinds like Job or JobSet that could have multiple Kueue-managed parent Kinds, each has to be covered by the check.
